src/adc.rs

```diff
@@ -27,358 +27,195 @@ static mut SPI_START: [u16; 1] = [0x00];
+
+                // The data transfer is always a transfer of data from the peripheral to a RAM buffer.
+                let mut data_transfer: Transfer<_, _, PeripheralToMemory, _> =
+                    Transfer::init(
+                        data_stream,
+                        spi,
+                        // Note(unsafe): The ADC_BUF[$index][0] is "owned" by this peripheral.
+                        // It shall not be used anywhere else in the module.
+                        unsafe { &mut ADC_BUF[$index][0] },
+                        None,
+                        data_config,
+                    );
+
+                data_transfer.start(|spi| {
+                    // Allow the SPI FIFOs to operate using only DMA data channels.
+                    spi.enable_dma_rx();
+                    spi.enable_dma_tx();
+
+                    // Enable SPI and start it in infinite transaction mode.
+                    spi.inner().cr1.modify(|_, w| w.spe().set_bit());
+                    spi.inner().cr1.modify(|_, w| w.cstart().started());
+                });
+
+                trigger_transfer.start(|_| {});
+
+                Self {
+                    // Note(unsafe): The ADC_BUF[$index][1] is "owned" by this peripheral. It shall not be used
+                    // anywhere else in the module.
+                    next_buffer: unsafe { Some(&mut ADC_BUF[$index][1]) },
+                    transfer: data_transfer,
+                    _trigger_transfer: trigger_transfer,
+                }
+            }
+
+            /// Obtain a buffer filled with ADC samples.
+            ///
+            /// # Returns
+            /// A reference to the underlying buffer that has been filled with ADC samples.
+            pub fn acquire_buffer(
+                &mut self,
+            ) -> &'static mut [u16; SAMPLE_BUFFER_SIZE] {
+                // Wait for the transfer to fully complete before continuing.
+                // Note: If a device hangs up, check that this conditional is passing correctly, as there is
+                // no time-out checks here in the interest of execution speed.
+                while !self.transfer.get_transfer_complete_flag() {}
+
+                let next_buffer = self.next_buffer.take().unwrap();
+
+                // Start the next transfer.
+                self.transfer.clear_interrupts();
+                let (prev_buffer, _) =
+                    self.transfer.next_transfer(next_buffer).unwrap();
+
+                prev_buffer
+            }
+
+            /// Release a buffer of ADC samples to the pool.
+            ///
+            /// # Args
+            /// * `next_buffer` - Buffer of ADC samples to be re-used.
+            pub fn release_buffer(
+                &mut self,
+                next_buffer: &'static mut [u16; SAMPLE_BUFFER_SIZE],
+            ) {
+                self.next_buffer.replace(next_buffer); // .unwrap_none() https://github.com/rust-lang/rust/issues/62633
```

Comment:
Why restructure this into two functions? We were previously leveraging rust ownership semantics to guarantee this via borrowing, but now we're incurring a run-time overhead? Is this to explicitly make it clear when the buffer is released as opposed to implicitly allowing the borrow to terminate due to scope?

Reply:
It was. And also to keep up the symmetry with the DAC API to help me understand things and differential design.
Let's revert that and see whether there are further changes around that DMA API that may warrant this.